Workflow validation

KevinMenden opened the pull request:
## PR checklist

 - [x] This comment contains a description of changes (with reason)
 - [x] `CHANGELOG.md` is updated
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [ ] Documentation in `docs` is updated

Added the lint function `actions_schema_validation` which uses the [GitHub workflow JSON schema](https://json.schemastore.org/github-workflow) to validate GitHub action workflows in `.github/workflows/`.

As suggested in #795

ewels commented on nf_core/lint/actions_schema_validation.py:
```diff
@@ -0,0 +1,53 @@
+    to verify that all workflow yml files adhere to the correct Schema
+    """
+    passed = []
+    warned = []
```
I don't think you have any warnings, so can remove this

ewels commented on nf_core/lint/actions_schema_validation.py:
```diff
@@ -0,0 +1,53 @@
+        except Exception as e:
+            failed.append("Workflow validation failed for {}: {}".format(wf, e))
+
+    return {"passed": passed, "warned": warned, "failed": failed}
```
No need to return warnings if there are none

ewels commented on nf_core/lint/actions_schema_validation.py:
```diff
@@ -0,0 +1,53 @@
+        # load workflow
+        wf_path = os.path.join(self.wf_path, ".github/workflows", wf)
+        with open(wf_path, "r") as fh:
+            wf_json = yaml.safe_load(fh)
```
Would be good to catch errors raised from yaml syntax problems and log as a failure here.

KevinMenden replied:
Good point - those files are also loaded in the other actions_* lint checks, I guess it makes sense to catch those errors there as well? Could basically just copy-paste the same code. We'll have a few redundant errors reported then, if the yml is wrong.

ewels commented on nf_core/lint/actions_schema_validation.py:
```diff
@@ -0,0 +1,53 @@
+    logging.getLogger("nf_core.schema").setLevel(logging.ERROR)
+
+    # Get all workflow files
+    action_workflows = os.listdir(os.path.join(self.wf_path, ".github/workflows"))
```
Maybe worth only getting `.yml` / `.yaml` files? In case any other auxiliary stuff is dumped there..

KevinMenden replied:
Yeah thought about this for a minute but then figured there should only be workflow files in there ... but you're right, better safe than sorry :)

ewels reviewed (commented):
Love it! Super clean code, just as I'd hoped 🚀 

Would be nice to add the ability to ignore specific files if listed in the linting config. Pattern for doing this in some other lint tests somewhere.

KevinMenden commented:
Thanks for the review!
Sure, should be easy to add to the code.

ewels commented on nf_core/lint/actions_schema_validation.py:
```diff
@@ -0,0 +1,58 @@
+            with open(wf_path, "r") as fh:
+                wf_json = yaml.safe_load(fh)
+        except Exception as e:
+            failed.append("Could not parse yaml file: {}".format(wf))
```
Might be nice to include `e` in the log to explain what went wrong.. (or a log debug if you want it short?)

KevinMenden replied:
Yep, was fearing that it get's too messy but can't hurt to have more information. Thanks!

ewels commented on nf_core/lint/actions_schema_validation.py:
```diff
@@ -0,0 +1,58 @@
+    Validate workflows against a schema
+
+    Uses the JSON Schema from
+    https://json.schemastore.org/github-workflow
```
Needs to be valid RST to make the hyperlink work in the docs.

Would also be good to add a bit more detail here - which files exactly? And where to get help (the URL is just a big JSON document). Have to assume that people reading this might not know that these files even exist..

KevinMenden replied:
Aaah sorry, didn't think about that this will be rendered at that moment 😱 
Will write a nicer message, thanks!

ewels reviewed (approved):
Love it! Haven't tested but review on my phone looks good. Couple of very minor suggestions but LGTM ✅

KevinMenden commented:
Added a longer description - not sure if there's anything else worth mentioning there except what these workflow files are for, where they are, and what the lint test checks. Let me know if something else comes to your mind!

Also added a commit to cast the exception at the schema validation to string, which should make the code more robust (as mentioned in slack).

ewels reviewed (approved):
Perfect ✨
